### PR TITLE
feat: [CDS-51715]: Api for moving inline template to remote template

### PR DESCRIPTION
**Pull request by utkarshchoubey-harness**
Api for moving inline template to remote template
 
 Before Template stpgrpp1 version 1 was inline as we can see it from the DB

image

After moving to remote it looks like this in DB :-
image

Api Response :- 
{
    "status": "SUCCESS",
    "data": {
        "templateIdentifier": "stpgrpp1",
        "versionLabel": "1"
    },
    "metaData": null,
    "correlationId": "56e7804c-341c-46cc-8a65-13b39083de03"
}

Tested scenerios with non-default branch as well and verified load-from-fallback branch is also working appropiately

## Checklist
- [x] I've documented the changes in the PR description.
- [x] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>

You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
    - CodeFormat: `trigger codeformat`
    - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/43488)

### CODE REVIEW

File: 970-ng-commons/src/main/java/io/harness/NGCommonEntityConstants.java [hunk 1]
```diff
@@ -47,6 +47,7 @@ public class NGCommonEntityConstants {
   public static final String DESCRIPTION_KEY = "description";
   public static final String DELETED_KEY = "deleted";
   public static final String PIPELINE_KEY = "pipelineIdentifier";
+  public static final String TEMPLATE_KEY = "templateIdentifier";
```
Comment (yogesh-chauhan) [inline 970-ng-commons/src/main/java/io/harness/NGCommonEntityConstants.java:50]: can we keep the template specific constants in Template Service if possible ?

**Review by yogesh-chauhan — CHANGES_REQUESTED** [review]

**Review by utkarshchoubey-harness — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/43488)* status: 13 file(s) to review. 1 change requests. 8. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`840-template-service/src/main/java/io/harness/template/beans/TemplateMoveConfigResponse.java` line 27 at r2](https://reviewable.io/reviews/harness/harness-core/43488#-NNM-R3Izqj88xpzskiQ-r2-27:-NNaSm5s6GVPXCCbfJ90:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/345536bfe307d335d960e0ee4aa1bd98f09d8bd0/840-template-service/src/main/java/io/harness/template/beans/TemplateMoveConfigResponse.java#L27)):*
<details><summary><i>Previously, yogesh-chauhan (Yogesh Chauhan) wrote…</i></summary><blockquote>

Tells us if move config -> Tells us if the move operation ?
</blockquote></details>

Done.
___
*[`840-template-service/src/main/java/io/harness/template/resources/NGTemplateResource.java` line 973 at r2](https://reviewable.io/reviews/harness/harness-core/43488#-NNM-R3Izqj88xpzskiG-r2-973:-NNaSmjc8gMqpOqdFPCY:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/345536bfe307d335d960e0ee4aa1bd98f09d8bd0/840-template-service/src/main/java/io/harness/template/resources/NGTemplateResource.java#L973)):*
<details><summary><i>Previously, yogesh-chauhan (Yogesh Chauhan) wrote…</i></summary><blockquote>

the move config contains the template identifier and version also, any reason these are duplicated as query params also ?
</blockquote></details>

Done.
___
*[`840-template-service/src/main/java/io/harness/template/services/NGTemplateServiceImpl.java` line 1345 at r2](https://reviewable.io/reviews/harness/harness-core/43488#-NNM-R3Izqj88xpzskiJ-r2-1345:-NNaaq8N0r-OBPkoBIqf:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/345536bfe307d335d960e0ee4aa1bd98f09d8bd0/840-template-service/src/main/java/io/harness/template/services/NGTemplateServiceImpl.java#L1345)):*
<details><summary><i>Previously, yogesh-chauhan (Yogesh Chauhan) wrote…</i></summary><blockquote>

the name of the method can be better. MoveConfig does not give much idea about what it will do
</blockquote></details>

Done.
___
*[`840-template-service/src/main/java/io/harness/template/services/NGTemplateServiceImpl.java` line 1348 at r2](https://reviewable.io/reviews/harness/harness-core/43488#-NNM-R3Izqj88xpzskiJ-r2-1348:-NNaSoe3A1fFhGSCfqZ6:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/345536bfe307d335d960e0ee4aa1bd98f09d8bd0/840-template-service/src/main/java/io/harness/template/services/NGTemplateServiceImpl.java#L1348)):*
<details><summary><i>Previously, yogesh-chauhan (Yogesh Chauhan) wrote…</i></summary><blockquote>

ideally we should not ask for the same information from 2 different places i.e query params and body
</blockquote></details>

Done.
___
*[`840-template-service/src/main/java/io/harness/template/services/NGTemplateServiceImpl.java` line 1354 at r2](https://reviewable.io/reviews/harness/harness-core/43488#-NNM-R3Izqj88xpzskiJ-r2-1354:-NNab3Cc9r6UqFGswjps:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/345536bfe307d335d960e0ee4aa1bd98f09d8bd0/840-template-service/src/main/java/io/harness/template/services/NGTemplateServiceImpl.java#L1354)):*
<details><summary><i>Previously, yogesh-chauhan (Yogesh Chauhan) wrote…</i></summary><blockquote>

lets not use fully qualified import for readability
</blockquote></details>

Done.
___
*[`840-template-service/src/main/java/io/harness/template/services/NGTemplateServiceImpl.java` line 1376 at r2](https://reviewable.io/reviews/harness/harness-core/43488#-NNM-R3Izqj88xpzskiJ-r2-1376:-NNahQXt8hmcwD1CTfI6:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/345536bfe307d335d960e0ee4aa1bd98f09d8bd0/840-template-service/src/main/java/io/harness/template/services/NGTemplateServiceImpl.java#L1376)):*
<details><summary><i>Previously, yogesh-chauhan (Yogesh Chauhan) wrote…</i></summary><blockquote>

shouldn't this be private ? Also we don't need to write tests for private method 

Lets write exhaustive tests of the above public method ```moveConfig```
</blockquote></details>

Done.
___
*[`840-template-service/src/main/java/io/harness/template/services/NGTemplateServiceImpl.java` line 1400 at r2](https://reviewable.io/reviews/harness/harness-core/43488#-NNM-R3Izqj88xpzskiJ-r2-1400:-NNabCnsE9mvVlfl1-Pr:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/345536bfe307d335d960e0ee4aa1bd98f09d8bd0/840-template-service/src/main/java/io/harness/template/services/NGTemplateServiceImpl.java#L1400)):*
<details><summary><i>Previously, yogesh-chauhan (Yogesh Chauhan) wrote…</i></summary><blockquote>

we should not log this error. Its an error by the user in the inputs to our API
Lets just throw an exception from here, it will be logged automatically at the REST layer
</blockquote></details>

Done.
___
*[`970-ng-commons/src/main/java/io/harness/NGCommonEntityConstants.java` line 50 at r2](https://reviewable.io/reviews/harness/harness-core/43488#-NNM-R3Izqj88xpzskiM-r2-50:-NNacWNb-obC9KlZD6rQ:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/345536bfe307d335d960e0ee4aa1bd98f09d8bd0/970-ng-commons/src/main/java/io/harness/NGCommonEntityConstants.java#L50)):*
<details><summary><i>Previously, yogesh-chauhan (Yogesh Chauhan) wrote…</i></summary><blockquote>

can we keep the template specific constants in Template Service if possible ?
</blockquote></details>

Done.

**Comment by yogesh-chauhan** [comment]
*[`840-template-service/src/main/java/io/harness/template/services/NGTemplateServiceImpl.java` line 1376 at r2](https://reviewable.io/reviews/harness/harness-core/43488#-NNM-R3Izqj88xpzskiJ-r2-1376:-NNbNcOf89ufduRkJF1F:b-84uxv1) ([raw file](https://github.com/harness/harness-core/blob/345536bfe307d335d960e0ee4aa1bd98f09d8bd0/840-template-service/src/main/java/io/harness/template/services/NGTemplateServiceImpl.java#L1376)):*
<details><summary><i>Previously, utkarshchoubey-harness wrote…</i></summary><blockquote>

Done.
</blockquote></details>

for tests, the method can be package private

**Review by yogesh-chauhan — APPROVED** [review]
> Reviewed 3 of 12 files at r1, 2 of 6 files at r2, 8 of 8 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/43488)* status: 13 file(s) to review. 1 of 2 approvals obtained.

**Review by tchaurasiya — DISMISSED** [review]
> Reviewed 3 of 12 files at r1, 2 of 6 files at r2, 8 of 8 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/43488)* status: :shipit: complete! ✓

**Comment by mohitg0795** [comment]
*[`840-template-service/src/main/java/io/harness/template/resources/NGTemplateResource.java` line 952 at r3](https://reviewable.io/reviews/harness/harness-core/43488#-NNecCow4o7G1hN3Y6zD:-NNecCow4o7G1hN3Y6zE:b1xvf7n) ([raw file](https://github.com/harness/harness-core/blob/aeeaf93a054a9a14bf64677542cf35ba6303f2db/840-template-service/src/main/java/io/harness/template/resources/NGTemplateResource.java#L952)):*
> ```Java
>   }
> 
>   @POST
> ```

This endpoint should be hidden from swagger, right?

**Comment by mohitg0795** [comment]
*[`840-template-service/src/main/java/io/harness/template/services/NGTemplateService.java` line 105 at r3](https://reviewable.io/reviews/harness/harness-core/43488#-NNec_X5AGa-IYy6fHEV:-NNec_X5AGa-IYy6fHEW:bhljx8o) ([raw file](https://github.com/harness/harness-core/blob/aeeaf93a054a9a14bf64677542cf35ba6303f2db/840-template-service/src/main/java/io/harness/template/services/NGTemplateService.java#L105)):*
> ```Java
> 
>   TemplateMoveConfigResponse moveTemplateStoreTypeConfig(String accountIdentifier, String orgIdentifier,
>       String projectIdentifier, String templateIdentifier, String versionLabel,
> ```

why is version label a separate param if already TemplateMoveConfigRequestDTO contains it?
Also, its fine for now, but ideally service layer request/response structs should be different from API layer structs. We shouldn't use API layer structs directly in service layer.

**Review by adivishy1 — APPROVED** [review]
> Reviewed 3 of 12 files at r1, 2 of 6 files at r2, 8 of 8 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/43488)* status: :shipit: complete! ✓

**Review by tchaurasiya — DISMISSED** [review]
> Reviewed 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/43488)* status: 1 file(s) to review. 1 of 2 approvals obtained.

**Comment by mohitg0795** [comment]
*[`840-template-service/src/main/java/io/harness/template/resources/NGTemplateResource.java` line 954 at r3](https://reviewable.io/reviews/harness/harness-core/43488#-NNehlg0CTPXySJMllNq:-NNehlg0CTPXySJMllNr:blnqmuh) ([raw file](https://github.com/harness/harness-core/blob/aeeaf93a054a9a14bf64677542cf35ba6303f2db/840-template-service/src/main/java/io/harness/template/resources/NGTemplateResource.java#L954)):*
> ```Java
>   @POST
>   @Path("/move-config/{templateIdentifier}")
>   @ApiOperation(value = "Move Template YAML from inline to remote", nickname = "moveTemplateConfigs")
> ```

its not only from inline to remote. It can be either ways.

**Review by adivishy1 — APPROVED** [review]
> Reviewed 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/43488)* status: :shipit: complete! ✓

**Comment by mohitg0795** [comment]
*[`840-template-service/src/test/java/io/harness/template/services/NGTemplateServiceImplTest.java` line 1161 at r3](https://reviewable.io/reviews/harness/harness-core/43488#-NNeiTs_8dKHPeb0abIR:-NNeiTs_8dKHPeb0abIS:blc16al) ([raw file](https://github.com/harness/harness-core/blob/aeeaf93a054a9a14bf64677542cf35ba6303f2db/840-template-service/src/test/java/io/harness/template/services/NGTemplateServiceImplTest.java#L1161)):*
> ```Java
>             .repoName("repo")
>             .branch("branch")
>             .moveConfigOperationType(io.harness.pms.pipeline.MoveConfigOperationType.INLINE_TO_REMOTE)
> ```

Why are we using pipeline struct here?

**Comment by mohitg0795** [comment]
*[`840-template-service/src/main/java/io/harness/template/services/NGTemplateServiceImpl.java` line 1427 at r3](https://reviewable.io/reviews/harness/harness-core/43488#-NNeltQV8SDxD2_wrd93:-NNeltQV8SDxD2_wrd94:bzazgak) ([raw file](https://github.com/harness/harness-core/blob/aeeaf93a054a9a14bf64677542cf35ba6303f2db/840-template-service/src/main/java/io/harness/template/services/NGTemplateServiceImpl.java#L1427)):*
> ```Java
>   }
> 
>   private ScmCreateFileGitResponse createRemoteEntity(TemplateEntity templateEntityToMove) {
> ```

templateEntityToMove is nullable here, we should have a check somewhere before to avoid it

**Review by utkarshchoubey-harness — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/43488)* status: 7 file(s) to review. 5. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`840-template-service/src/main/java/io/harness/template/resources/NGTemplateResource.java` line 952 at r3](https://reviewable.io/reviews/harness/harness-core/43488#-NNecCow4o7G1hN3Y6zD:-NNf13fc1o6R0J6a-ucR:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/aeeaf93a054a9a14bf64677542cf35ba6303f2db/840-template-service/src/main/java/io/harness/template/resources/NGTemplateResource.java#L952)):*
<details><summary><i>Previously, mohitg0795 (Mohit Garg) wrote…</i></summary><blockquote>

This endpoint should be hidden from swagger, right?
</blockquote></details>

Done.
___
*[`840-template-service/src/main/java/io/harness/template/resources/NGTemplateResource.java` line 954 at r3](https://reviewable.io/reviews/harness/harness-core/43488#-NNehlg0CTPXySJMllNq:-NNf13xj-pAhjVvpXMGT:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/aeeaf93a054a9a14bf64677542cf35ba6303f2db/840-template-service/src/main/java/io/harness/template/resources/NGTemplateResource.java#L954)):*
<details><summary><i>Previously, mohitg0795 (Mohit Garg) wrote…</i></summary><blockquote>

its not only from inline to remote. It can be either ways.
</blockquote></details>

Done.
___
*[`840-template-service/src/main/java/io/harness/template/services/NGTemplateService.java` line 105 at r3](https://reviewable.io/reviews/harness/harness-core/43488#-NNec_X5AGa-IYy6fHEV:-NNf14iI9IAtkoE8_U1B:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/aeeaf93a054a9a14bf64677542cf35ba6303f2db/840-template-service/src/main/java/io/harness/template/services/NGTemplateService.java#L105)):*
<details><summary><i>Previously, mohitg0795 (Mohit Garg) wrote…</i></summary><blockquote>

why is version label a separate param if already TemplateMoveConfigRequestDTO contains it?
Also, its fine for now, but ideally service layer request/response structs should be different from API layer structs. We shouldn't use API layer structs directly in service layer.
</blockquote></details>

Done.
___
*[`840-template-service/src/main/java/io/harness/template/services/NGTemplateServiceImpl.java` line 1427 at r3](https://reviewable.io/reviews/harness/harness-core/43488#-NNeltQV8SDxD2_wrd93:-NNf18EJ0XA3bJbLLD4d:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/aeeaf93a054a9a14bf64677542cf35ba6303f2db/840-template-service/src/main/java/io/harness/template/services/NGTemplateServiceImpl.java#L1427)):*
<details><summary><i>Previously, mohitg0795 (Mohit Garg) wrote…</i></summary><blockquote>

templateEntityToMove is nullable here, we should have a check somewhere before to avoid it
</blockquote></details>

Done.
___
*[`840-template-service/src/test/java/io/harness/template/services/NGTemplateServiceImplTest.java` line 1161 at r3](https://reviewable.io/reviews/harness/harness-core/43488#-NNeiTs_8dKHPeb0abIR:-NNf19-M9caEd0Mt8X3g:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/aeeaf93a054a9a14bf64677542cf35ba6303f2db/840-template-service/src/test/java/io/harness/template/services/NGTemplateServiceImplTest.java#L1161)):*
<details><summary><i>Previously, mohitg0795 (Mohit Garg) wrote…</i></summary><blockquote>

Why are we using pipeline struct here?
</blockquote></details>

Done.

**Comment by utkarshchoubey-harness** [comment]
trigger ti0

**Comment by mohitg0795** [comment]
*[`840-template-service/src/main/java/io/harness/template/services/NGTemplateService.java` line 105 at r3](https://reviewable.io/reviews/harness/harness-core/43488#-NNec_X5AGa-IYy6fHEV:-NNf5jJT47URGBmSyNWY:by3kkjx) ([raw file](https://github.com/harness/harness-core/blob/aeeaf93a054a9a14bf64677542cf35ba6303f2db/840-template-service/src/main/java/io/harness/template/services/NGTemplateService.java#L105)):*
<details><summary><i>Previously, utkarshchoubey-harness wrote…</i></summary><blockquote>

Done.
</blockquote></details>

what is the change here?

**Comment by mohitg0795** [comment]
*[`840-template-service/src/main/java/io/harness/template/services/NGTemplateServiceImpl.java` line 1374 at r5](https://reviewable.io/reviews/harness/harness-core/43488#-NNf5qfy7yNgwoksiHYS:-NNf5qfy7yNgwoksiHYT:bp7qyww) ([raw file](https://github.com/harness/harness-core/blob/1b1f76260e1967fc6c56d8af1b0f5e35d45de4cb/840-template-service/src/main/java/io/harness/template/services/NGTemplateServiceImpl.java#L1374)):*
> ```Java
>     } else {
>       throw new InvalidRequestException(
>           String.format("Error while retrieving template with identifier [%s] and versionLabel [%s]",
> ```

we should say that template with given id is not found, right?

**Review by utkarshchoubey-harness — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/43488)* status: 9 file(s) to review. 2. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`840-template-service/src/main/java/io/harness/template/services/NGTemplateService.java` line 105 at r3](https://reviewable.io/reviews/harness/harness-core/43488#-NNec_X5AGa-IYy6fHEV:-NNf9AJ8C-R-4VIbbGB_:b-7qwb9r) ([raw file](https://github.com/harness/harness-core/blob/aeeaf93a054a9a14bf64677542cf35ba6303f2db/840-template-service/src/main/java/io/harness/template/services/NGTemplateService.java#L105)):*
<details><summary><i>Previously, mohitg0795 (Mohit Garg) wrote…</i></summary><blockquote>

what is the change here?
</blockquote></details>

Done  can you check again once
___
*[`840-template-service/src/main/java/io/harness/template/services/NGTemplateServiceImpl.java` line 1374 at r5](https://reviewable.io/reviews/harness/harness-core/43488#-NNf5qfy7yNgwoksiHYS:-NNf9EAt-TzCr2ElFZUH:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/1b1f76260e1967fc6c56d8af1b0f5e35d45de4cb/840-template-service/src/main/java/io/harness/template/services/NGTemplateServiceImpl.java#L1374)):*
<details><summary><i>Previously, mohitg0795 (Mohit Garg) wrote…</i></summary><blockquote>

we should say that template with given id is not found, right?
</blockquote></details>

Done.

**Review by mohitg0795 — APPROVED** [review]
> Reviewed 2 of 12 files at r1, 2 of 6 files at r2, 2 of 8 files at r3, 5 of 7 files at r5, 4 of 4 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/43488)* status: 9 file(s) to review. 1 of 2 approvals obtained.

**Review by yogesh-chauhan — APPROVED** [review]
> Reviewed 5 of 7 files at r5, 4 of 4 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/43488)* status: :shipit: complete! ✓

**Review by tchaurasiya — APPROVED** [review]
> Reviewed 5 of 7 files at r5, 4 of 4 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/43488)* status: :shipit: complete! ✓

**Review by abhinavmittal-harness — APPROVED** [review]
> Reviewed 2 of 12 files at r1, 2 of 6 files at r2, 2 of 8 files at r3, 5 of 7 files at r5, 4 of 4 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/43488)* status: :shipit: complete! ✓

**Pull request merged**